### PR TITLE
[Composable] Use non-reentrant generator, remove reentrant

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106003
* #106002
* #106001
* #106000
* #105999
* #105995
* #105991
* #105207
* #105206
* #105205
* #105180
* __->__ #105176

Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

### CODE REVIEW

File: torch/distributed/_composable/checkpoint_activation.py
Comment (awgu) [inline torch/distributed/_composable/checkpoint_activation.py:99]: Given that we are not passing kwargs into the hooks, I think we will still need to fix for supporting kwargs.

Comment (rohan-varma) [reply torch/distributed/_composable/checkpoint_activation.py:99]: sounds good, we can fix this in the follow up diffs?

File: test/distributed/_composable/test_checkpoint.py
Comment (awgu) [inline test/distributed/_composable/test_checkpoint.py:11]: Remove the `use_reentrant` kwarg in `test_checkpoint_fsdp_submodules_non_reentrant()`:
```
test_model.u1.seq = checkpoint(test_model.u1.seq, use_reentrant=use_reentrant)
```

File: test/distributed/_composable/test_checkpoint.py
Comment (awgu) [inline test/distributed/_composable/test_checkpoint.py:211]:
```suggestion
        with self.assertRaisesRegex(RuntimeError, "foo"):
```
(or factor the runtime error message into a local variable and use that)

File: test/distributed/_composable/test_checkpoint.py
Comment (awgu) [inline test/distributed/_composable/test_checkpoint.py:188]: nit: For AC specifically, it can be good to think about applying it to a 'module' as opposed to a 'model' because realistically we would never checkpoint the entire model.

File: test/distributed/_composable/test_checkpoint.py
Comment (awgu) [inline test/distributed/_composable/test_checkpoint.py:186]: nit: Maybe we could call this test `test_clears_state_on_error_in_forward()` (or similar) or somehow otherwise clarify that this is testing that `checkpoint`-specific state is cleared on an error in forward?

File: torch/distributed/_composable/checkpoint_activation.py
Comment (awgu) [inline torch/distributed/_composable/checkpoint_activation.py:89]: What happens if we hit this error? What do we do?

**Review by awgu — APPROVED** [review]
Stamp to unblock -- please fix the failing unit test and lint before landing :)

File: torch/distributed/_composable/checkpoint_activation.py
Comment (rohan-varma) [reply torch/distributed/_composable/checkpoint_activation.py:89]: I'm not sure if this can happen other than an implementation change / bug where the AC generator can be `next`'ed more than twice. I guess we should tell the user to file a repro?

Comment (awgu) [reply torch/distributed/_composable/checkpoint_activation.py:89]: Sounds good!

**Review by fegin — APPROVED** [review]

File: test/distributed/_composable/test_checkpoint.py
Comment (rohan-varma) [reply test/distributed/_composable/test_checkpoint.py:184]: sounds good, will refactor this into a container

**Pull request closed without merging**
